[maps][ios][android] add Polyline support

fobos531 opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

I found myself needing to have support for polylines. Open to suggestions regarding API design.

By the way, I believe a big plus to support in the future would be rendering arbitrary React views as markers, but I'm not sure how compatible that is with the current approach of rendering (where we render markers/annotations as arrays, instead of how e.g. `react-native-maps` does it with such elements being children of the MapView)

# How

<!--
How did you build this feature or fix this bug and why?
-->

Same approach as markers.

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

There is now a polyline screen.

![CleanShot 2025-04-17 at 17 11 47@2x](https://github.com/user-attachments/assets/6cb7df4d-c4ff-4403-b4cb-012ba6f0894c)

![CleanShot 2025-04-17 at 17 12 10@2x](https://github.com/user-attachments/assets/6cbf0e4d-307e-4711-b624-52bf6707948e)


# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [X] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

fobos531 committed:
[maps][ios][android] add polyline support

fobos531 committed:
add changelog entry

fobos531 committed:
run swiftlint

fobos531 committed:
swiftlint fixes

fobos531 committed:
swiftlint fixes

fobos531 committed:
cleanup

fobos531 committed:
fix references

fobos531 committed:
undo irrelevant changes & swiftlint fixes

fobos531 committed:
undo more irrelevant changes

fobos531 committed:
swiftlint tweaks

fobos531 commented on packages/expo-maps/ios/AppleMapsView.swift:
```diff
@@ -159,4 +198,43 @@ struct AppleMapsView: View, AppleMapsViewProtocol {
+
+  private func polyline(
+    at tap: CLLocationCoordinate2D,
+    threshold: CLLocationDistance = 20
```
since apple maps doesnt support tapping on polylines by default, this is a custom solution to allow tapping on them. threshold is arbitrary here, perhaps it should be configured by the user? kinda like `hitSlop` for touchables

lukmccall replied:
Let's add an option for the user to configure it. We don't want to hide it.

fobos531 committed:
fix swiftlint

fobos531 commented on packages/expo-maps/ios/AppleMapsView.swift:
```diff
@@ -122,19 +173,7 @@ struct AppleMapsView: View, AppleMapsViewProtocol {
-          ])
-        }
-      }
+      .onChange(of: state.selection, perform: handleSelectionChange)
```
this was needed because xcode complained about the compiler not being able to type-check in reasonable time

lukmccall commented on apps/native-component-list/src/screens/ExpoMaps/apple/MapsPolylineScreen.tsx:
```diff
@@ -0,0 +1,371 @@
+          console.log(event);
+          Alert.alert('Polyline clicked', JSON.stringify(event));
+        }}
+        polylines={[
+          {
+            strokeColor: '#ff0000',
+            strokeWidth: 20,
+            coordinates: [
+              {
+                latitude: 46.334295,
+                longitude: 16.5656,
+              },
+              {
+                latitude: 46.334278,
+                longitude: 16.565874,
+              },
+              {
```
Can we reuse that data between the iOS and Android implementations? Also, let's extract it to a separate file. It's making the code less readable right now.

fobos531 replied:
Sure, extracted and reused.

lukmccall reviewed (commented):
At first glance, it looks good. I'll do a second pass soon. Thanks for your contribution!

fobos531 committed:
make code more readable

fobos531 commented:
To add to my comment about custom Marker support - in a follow up PR we might want to switch to a more composable API, like we did for ContextMenu. I believe the reasoning is very similar. Thoughts?

fobos531 committed:
lint

lukmccall commented on packages/expo-maps/ios/AppleMapsView.swift:
```diff
@@ -1,20 +1,46 @@
 }
 
+extension MKMapPoint {
+  /// Perpendicular distance (in metres) from `self` to the
+  /// line segment **AB**.
```
```suggestion
  // Perpendicular distance (in metres) from `self` to the
  // line segment **AB**.
```

lukmccall commented on packages/expo-maps/ios/AppleMapsView.swift:
```diff
@@ -1,20 +1,46 @@
+    // Clamp the projection to the segment
+    let clamped = max(0.0, min(1.0, t))
+
+    let proj = MKMapPoint(
+      x: a.x + clamped * dx,
+      y: a.y + clamped * dy)
```
```suggestion
    let proj = MKMapPoint(
      x: a.x + clamped * dx,
      y: a.y + clamped * dy
    )
```

lukmccall commented on packages/expo-maps/android/src/main/java/expo/modules/maps/Records.kt:
```diff
@@ -64,6 +64,20 @@ data class MarkerRecord(
+  val geodesic: Boolean = false,
+
+  @Field
+  val color: String = "#000000",
```
Color should be handled differently to ensure compatibility with strings and other formats.

For example, you can see how it's implemented in `expo-image`, where the color is represented as an `Int` on the native side (see https://github.com/expo/expo/blob/32f07127438a21f288aa7f3334881fb70b15064c/packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt#L226 and https://github.com/expo/expo/blob/32f07127438a21f288aa7f3334881fb70b15064c/packages/expo-image/src/ExpoImage.tsx#L80).

lukmccall commented:
> To add to my comment about custom Marker support - in a follow up PR we might want to switch to a more composable API for Maps, like we did for ContextMenu. I believe the reasoning is very similar. Thoughts?

That sounds appealing, but I believe we can support both options. The current one should perform better, especially for large lists of points.

fobos531 committed:
Apply suggestions from code review

Co-authored-by: Łukasz Kosmaty <kosmatylukasz@gmail.com>

fobos531 committed:
apply PR review feedback

fobos531 committed:
small google map fixes

fobos531 commented:
@lukmccall I have addressed your feedback. Polyline threshold config has been added to the map `properties`, **not** the polyline itself - I figured users will want to have the same threshold for all polylines, but if needed I can move it to polyline configuration itself. Let me know if you think any further changes to the API are needed.

> That sounds appealing, but I believe we can support both options. The current one should perform better, especially for large lists of points.

So both the current (props) and new (composition) approaches would be valid? This is fine with me - generally, props would be used when the use-case is simple, and when something more customized needs to happen (e.g. custom annotations/markers etc.) the composition can be used. 

Can you clarify why the current approach is better in terms of performance?

lukmccall commented on packages/expo-maps/src/google/GoogleMapsView.tsx:
```diff
@@ -56,6 +58,12 @@ export const GoogleMapsView = React.forwardRef<GoogleMapsViewType, GoogleMapsVie
+
+    const parsedPolylines = polylines?.map((polyline) => ({
+      ...polyline,
+      color: polyline.color ? processColor(polyline.color) : undefined,
```
```suggestion
      color: processColor(polyline.color),
```
I don't think that check is necessary.

fobos531 replied:
I'd love to remove it, but at the very least what we need is `color: processColor(polyline.color) ?? undefined` since processColor can return null, which causes typescript to complain.

lukmccall commented on packages/expo-maps/src/apple/AppleMapsView.tsx:
```diff
@@ -36,6 +39,12 @@ export const AppleMapsView = React.forwardRef<AppleMapsViewType, AppleMapsViewPr
+
+    const parsedPolylines = polylines?.map((polyline) => ({
+      ...polyline,
+      strokeColor: polyline.strokeColor ? processColor(polyline.strokeColor) : undefined,
```
```suggestion
      strokeColor: processColor(polyline.strokeColor),
```

lukmccall commented on packages/expo-maps/src/apple/AppleMaps.types.ts:
```diff
@@ -116,6 +139,25 @@ export type AppleMapsAnnotation = {
+  /**
+   * The color of the polyline.
+   */
+  strokeColor?: ProcessedColorValue | string | null;
+  /**
+   * The width of the polyline.
+   */
+  strokeWidth?: number;
```
```suggestion
  color?: ProcessedColorValue | string | null;
  /**
   * The width of the polyline.
   */
  width?: number;
```
I think we can use the same names on both platforms.

fobos531 replied:
Done

lukmccall commented:
> Polyline threshold config has been added to the map properties, not the polyline itself - I figured users will want to have the same threshold for all polylines 

That's okay, we don't need to move it to polylines.

> Can you clarify why the current approach is better in terms of performance?

Currently, we don't need to convert child components into an array that can be sent to the native side. This reduces the operations required by JavaScript.

fobos531 committed:
apply PR feedback

fobos531 commented:
@lukmccall Applied latest request for changes.

fobos531 committed:
Merge branch 'main' into @fobos531/polyline

fobos531 committed:
polyline updates

fobos531 committed:
run swiftlint

fobos531 committed:
lint

csulit commented:
👀

lukmccall reviewed (approved):
LGTM!
